stem/descriptor/hidden_service.py

```diff
@@ -0,0 +1,1350 @@
+    return hashlib.sha3_256(b'subcredential%s%s' % (credential, blinded_key)).digest()
+
+
+class OuterLayer(Descriptor):
+  """
+  Initial encryped layer of a hidden service v3 descriptor (`spec
+  <https://gitweb.torproject.org/torspec.git/tree/rend-spec-v3.txt#n1154>`_).
+
+  .. versionadded:: 1.8.0
+
+  :var str auth_type: **\\*** encryption scheme used for descriptor authorization
+  :var str ephemeral_key: **\\*** base64 encoded x25519 public key
+  :var dict clients: **\\*** mapping of authorized client ids to their
+    :class:`~stem.descriptor.hidden_service.AuthorizedClient`
+  :var str encrypted: **\\*** encrypted descriptor inner layer
+
+  **\\*** attribute is either required when we're parsed with validation or has
+  a default value, others are left as **None** if undefined
+  """
+
+  ATTRIBUTES = {
+    'auth_type': (None, _parse_v3_outer_auth_type),
+    'ephemeral_key': (None, _parse_v3_outer_ephemeral_key),
+    'clients': ({}, _parse_v3_outer_clients),
+    'encrypted': (None, _parse_v3_outer_encrypted),
+  }
+
+  PARSER_FOR_LINE = {
+    'desc-auth-type': _parse_v3_outer_auth_type,
+    'desc-auth-ephemeral-key': _parse_v3_outer_ephemeral_key,
+    'auth-client': _parse_v3_outer_clients,
+    'encrypted': _parse_v3_outer_encrypted,
+  }
+
+  @staticmethod
+  def _decrypt(encrypted, revision_counter, subcredential, blinded_key):
+    plaintext = _decrypt_layer(encrypted, b'hsdir-superencrypted-data', revision_counter, subcredential, blinded_key)
+    return OuterLayer(plaintext)
+
+  def _encrypt(self, revision_counter, subcredential, blinded_key):
+    # Spec mandated padding: "Before encryption the plaintext is padded with
+    # NUL bytes to the nearest multiple of 10k bytes."
+
+    content = self.get_bytes() + b'\x00' * (len(self.get_bytes()) % 10000)
+
+    # encrypt back into a hidden service descriptor's 'superencrypted' field
+
+    return _encrypt_layer(content, b'hsdir-superencrypted-data', revision_counter, subcredential, blinded_key)
+
+  @classmethod
+  def content(cls, attr = None, exclude = (), validate = True, sign = False, inner_layer = None, revision_counter = None, subcredential = None, blinded_key = None):
+    if not stem.prereq.is_crypto_available(ed25519 = True):
+      raise ImportError('Hidden service layer creation requires cryptography version 2.6')
+
+    from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
+
+    inner_layer = inner_layer if inner_layer else InnerLayer.create()
+    revision_counter = revision_counter if revision_counter else 1
+    blinded_key = blinded_key if blinded_key else stem.util._pubkey_bytes(Ed25519PrivateKey.generate())
+    subcredential = subcredential if subcredential else HiddenServiceDescriptorV3._subcredential(Ed25519PrivateKey.generate(), blinded_key)
+
+    return _descriptor_content(attr, exclude, (
+      ('desc-auth-type', 'x25519'),
+      ('desc-auth-ephemeral-key', base64.b64encode(os.urandom(32))),
```

Comment:
One of the goals of my old `_get_middle_descriptor_layer_body()` function here was to make the output of stem indistinguishable from the output of Tor. By using random bytes instead of an actual x25519 key here, this becomes a fingerprint that the descriptor is created by stem. Any chance you would consider reinstalling the functionality of `_get_middle_descriptor_layer_body()` for the purposes of encoding fake client auth data?

Remaining indistinguishable from Tor is an important security goal for the future. Mike Perry has spent lots of time thinking about this and asked me that we should support this in the future, so that's why I'm mentioning it here. Also see https://github.com/mikeperry-tor/vanguards/blob/master/README_SECURITY.md#adversaries-client for more information.